// @adobe/commerce-cif-model/index.js
/*******************************************************************************
 *
 *    Copyright 2018 Adobe. All rights reserved.
 *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License. You may obtain a copy
 *    of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software distributed under
 *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 *    OF ANY KIND, either express or implied. See the License for the specific language
 *    governing permissions and limitations under the License.
 *
 ******************************************************************************/

/**
 * Auto generated code based on Swagger definition.
 * Dot not edit manually. Manual changes will be overridden.
 *
 * @version 1.4.0
 */

module.exports.Address = require('./Address.js').Address;
module.exports.AddressWrapper = require('./AddressWrapper.js').AddressWrapper;
module.exports.Asset = require('./Asset.js').Asset;
module.exports.Attribute = require('./Attribute.js').Attribute;
module.exports.AuthenticationResponse = require('./AuthenticationResponse.js').AuthenticationResponse;
module.exports.Cart = require('./Cart.js').Cart;
module.exports.CartEntry = require('./CartEntry.js').CartEntry;
module.exports.Category = require('./Category.js').Category;
module.exports.Coupon = require('./Coupon.js').Coupon;
module.exports.Customer = require('./Customer.js').Customer;
module.exports.Discount = require('./Discount.js').Discount;
module.exports.ErrorResponse = require('./ErrorResponse.js').ErrorResponse;
module.exports.Facet = require('./Facet.js').Facet;
module.exports.FacetValue = require('./FacetValue.js').FacetValue;
module.exports.GraphqlRequest = require('./GraphqlRequest.js').GraphqlRequest;
module.exports.GraphqlResponse = require('./GraphqlResponse.js').GraphqlResponse;
module.exports.HealthResponse = require('./HealthResponse.js').HealthResponse;
module.exports.HealthResponseStatusReport = require('./HealthResponseStatusReport.js').HealthResponseStatusReport;
module.exports.InventoryItem = require('./InventoryItem.js').InventoryItem;
module.exports.LoginResult = require('./LoginResult.js').LoginResult;
module.exports.MoneyValue = require('./MoneyValue.js').MoneyValue;
module.exports.Order = require('./Order.js').Order;
module.exports.PagedResponse = require('./PagedResponse.js').PagedResponse;
module.exports.PagedResponseCategory = require('./PagedResponseCategory.js').PagedResponseCategory;
module.exports.PagedResponseInventoryItem = require('./PagedResponseInventoryItem.js').PagedResponseInventoryItem;
module.exports.PagedResponseProduct = require('./PagedResponseProduct.js').PagedResponseProduct;
module.exports.PagedResponseShoppingList = require('./PagedResponseShoppingList.js').PagedResponseShoppingList;
module.exports.PagedResponseShoppingListEntry = require('./PagedResponseShoppingListEntry.js').PagedResponseShoppingListEntry;
module.exports.Payment = require('./Payment.js').Payment;
module.exports.PaymentMethod = require('./PaymentMethod.js').PaymentMethod;
module.exports.PaymentWrapper = require('./PaymentWrapper.js').PaymentWrapper;
module.exports.Product = require('./Product.js').Product;
module.exports.ProductVariant = require('./ProductVariant.js').ProductVariant;
module.exports.ShippingInfo = require('./ShippingInfo.js').ShippingInfo;
module.exports.ShippingMethod = require('./ShippingMethod.js').ShippingMethod;
module.exports.ShoppingList = require('./ShoppingList.js').ShoppingList;
module.exports.ShoppingListEntry = require('./ShoppingListEntry.js').ShoppingListEntry;
module.exports.StatusReport = require('./StatusReport.js').StatusReport;
module.exports.TaxInfo = require('./TaxInfo.js').TaxInfo;
module.exports.TaxPortion = require('./TaxPortion.js').TaxPortion;
module.exports.swagger = require('./swagger.json');
module.exports.graphqlSchema = require('./graphqlSchema.js').schema;
